README.md
# @suin/gatsby-source-esa

Gatsby に esa.io のデータを提供するソースプラグイン。esa API v1 を使用してデータを取り込みます。

現在は、投稿の取得のみ対応しています。

## 特徴

- esa のデータソースに対応
  - 記事
- TypeScript サポート
  - GraphQL の[スキーマを定義している](./lib/createSchemaCustomization.ts)ので、[gatsby-plugin-graphql-codegen]が生成する TypeScript 型定義がしっかりしています。
- 複数の esa チームをデータソースに指定できる
- 下記の gatsby プラグインとの統合をサポート
  - gatsby-transformer-remark
  - gatsby-transformer-rehype

[gatsby-plugin-graphql-codegen]: https://www.gatsbyjs.org/packages/gatsby-plugin-graphql-codegen/

## インストール

```bash
yarn add @suin/gatsby-source-esa
# or
npm install @suin/gatsby-source-esa
```

## 使い方

### 設定の仕方

gatsby-config.js のプラグイン設定に@suin/gatsby-source-esa を追加してください:

```javascript
// gatsby-config.js
module.exports = {
  // ...
  plugins: [
    {
      resolve: `@suin/gatsby-source-esa`,
      options: {
        team: `foo`,
        token: process.env.ESA_TOKEN,
        posts: {
          q: `wip:false in:Public`,
          include: [`comments`, `stargazers`],
          sort: `number`,
          order: `asc`,
        },
      },
    },
    // ...
  ],
}
```

**オプション**

- `team`: https://{チーム名}.esa.io の {チーム名} の部分を指定する。
- `token`: ユーザの管理画面(https://[team].esa.io/user/tokens)で発行したアクセストークンを指定する。
- `posts`: エンドポイント [GET /v1/team/:team_name/posts の URI クエリ文字列](https://docs.esa.io/posts/102#URI%E3%82%AF%E3%82%A8%E3%83%AA%E6%96%87%E5%AD%97%E5%88%97-1) に指定できるパラメータが利用可能。
  - `posts.q`: 記事を絞り込むための条件。ここに検索オプション(wip:false や in:Public など)を駆使して、公開すべき記事をうまく絞り込むことをおすすめをします。[help/記事の検索方法 - docs.esa.io](https://docs.esa.io/posts/104)参照。

### クエリー

GraphQL でのクエリーは以下のように行なえます:

```graphql
query MyQuery {
  allEsaPost {
    edges {
      node {
        number
        name
        body_md
        body_html
        category
        tags
        created_at
        updated_at
      }
    }
  }
}
```

**提供される GraphQL インターフェイス**

- allEsaPost: すべての投稿
- allEsaPostBodyHtml: すべての投稿の本文の HTML 部分
- allEsaPostBodyMarkdown: すべての投稿の本文の Markdown 部分
- esaPost
- esaPostBodyHtml
- esaPostBodyMarkdown

## 注意事項

- esa API には「ユーザ毎に 15 分間に 75 リクエストまで」という利用制限があります。このプラグインはできるだけリクエスト数が少なくなるように設計してありますが、この利用制限に達した場合の挙動は未定義です。

## Tips

### gatsby-transformer-remark との統合

Markdown ファイルをパースする Gatsby プラグインの [gatsby-transformer-remark](https://www.gatsbyjs.org/packages/gatsby-transformer-remark/) とは自動的に統合されます。下記のように、当プラグインと gatsby-transformer-remark を併用するだけです:

```javascript
// gatsby-config.js
module.exports = {
  // ...
  plugins: [
    {
      resolve: `@suin/gatsby-source-esa`,
      options: {
        team: `foo`,
        token: process.env.ESA_TOKEN,
        posts: {
          /*...*/
        },
      },
    },
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        /*...*/
      },
    },
    // ...
  ],
}
```

### gatsby-transformer-rehype との統合

HTML ファイル(mediaType が text/html)のノードを変形する Gatsby プラグインの [gatsby-transformer-rehype](https://www.gatsbyjs.org/packages/gatsby-transformer-rehype/) とは自動的に統合されます。下記のように、当プラグインと gatsby-transformer-rehype を併用するだけです。

gatsby-transformer-rehype には記事ノードの`body_html`が渡ります。

```javascript
// gatsby-config.js
module.exports = {
  // ...
  plugins: [
    {
      resolve: `@suin/gatsby-source-esa`,
      options: {
        team: `foo`,
        token: process.env.ESA_TOKEN,
        posts: {
          /*...*/
        },
      },
    },
    {
      resolve: `gatsby-transformer-rehype`,
      options: {
        /*...*/
      },
    },
    // ...
  ],
}
```

gatsby-transformer-rehype で処理された HTML を取り出すには、次のようなクエリーを発行します:

```graphql
query MyQuery {
  allEsaPost {
    edges {
      node {
        number
        name
        childrenEsaPostBodyHtml {
          childHtmlRehype {
            html
          }
        }
      }
    }
  }
}
```

結果の例:

```json
{
  "data": {
    "allEsaPost": {
      "edges": [
        {
          "node": {
            "number": 123,
            "name": "...",
            "childrenEsaPostBodyHtml": [
              {
                "childHtmlRehype": {
                  "html": "<p>..."
                }
              }
            ]
          }
        }
      ]
    }
  }
}
```

### 複数のチームをデータソースにする場合

このプラグインは複数の esa チームをデータソースにしても、それぞれのチームの投稿が Gatsby 上で衝突しないように設計されています。

複数のチームからデータを取り込む場合は、下記のように`@suin/gatsby-source-esa`を複数記述してください。

```javascript
// gatsby-config.js
module.exports = {
  // ...
  plugins: [
    {
      resolve: `@suin/gatsby-source-esa`,
      options: {
        team: `foo`,
        token: process.env.ESA_FOO_TOKEN,
      },
    },
    {
      resolve: `@suin/gatsby-source-esa`,
      options: {
        team: `bar`,
        token: process.env.ESA_BAR_TOKEN,
      },
    },
    // ...
  ],
}
```

GraphQL でチーム名を参照するには、`team`をクエリに含めてください:

```graphql
query MyQuery {
  allEsaPost {
    edges {
      node {
        team # ここ
        name
        body_md
        # ....
      }
    }
  }
}
```


CHANGELOG.md
# [1.2.0](https://github.com/suin/gatsby-source-esa/compare/v1.1.0...v1.2.0) (2020-06-11)


### Features

* 🎸 export type `EsaPost` ([7df4919](https://github.com/suin/gatsby-source-esa/commit/7df4919e5fd4cf3077d1d41cc2050b1f7e8b17a5))
* 🎸 provide GraphQL schema to Gatsby ([32e5ed9](https://github.com/suin/gatsby-source-esa/commit/32e5ed9391ae517e70ca3cad4dea7d66c788f933))

# [1.1.0](https://github.com/suin/gatsby-source-esa/compare/v1.0.0...v1.1.0) (2020-05-24)


### Features

* 🎸 support gatsby-transformer-rehype ([0028489](https://github.com/suin/gatsby-source-esa/commit/002848906c297f7efbb68a51ba1ac9b5101ea981))

# 1.0.0 (2020-05-21)


### Features

* 🎸 implement it! ([e4a14a9](https://github.com/suin/gatsby-source-esa/commit/e4a14a962bd65c5f76e6487a5fc0dec2432a73ca))

# 1.0.0 (2020-05-21)


### Features

* 🎸 add implementation ([3a78965](https://github.com/suin/gatsby-source-esa/commit/3a78965530d08af4a9771f206b82bff4148bcb48))


lib/errors.test.ts
import { Errors } from './errors'

describe('errors', () => {
  test('example', () => {
    const errors = new Errors()
    expect(errors.noErrors()).toBe(true)
    errors.add('error1')
    errors.add('error2', 'error3')
    expect(errors.noErrors()).toBe(false)
    expect(errors.toString()).toBe('error1\nerror2\nerror3')
  })
})


lib/esaApi.test.ts
import MockAdapter from 'axios-mock-adapter'
import { EsaApi, PostsPayload, Post } from './esaApi'

describe('esaApi', () => {
  let api: EsaApi
  let mock: MockAdapter

  beforeEach(() => {
    api = new EsaApi({
      team: 'acme',
      token: 'mock-token',
    })
    mock = new MockAdapter(api.client)
  })

  describe('integration testing', () => {
    const { ESA_TEAM: team, ESA_TOKEN: token } = process.env as any
    if (typeof team !== 'string' || typeof token !== 'string') {
      console.warn(
        'skipped integration testing since `ESA_TEAM` and `ESA_TOKEN` was not provided',
      )
      return
    }
    test('get one post', async () => {
      api = new EsaApi({ team, token })
      const { posts } = await api.getPosts({ per_page: 1 })
      expect(posts.length).toBe(1)
    })
  })

  describe('enhead authorization', () => {
    test('token', async () => {
      mock.onGet('/teams/acme/posts').replyOnce(({ headers }) => {
        expect(headers).toMatchObject({ Authorization: 'Bearer mock-token' })
        return [200, {}, {}]
      })
      await api.getPosts()
    })
  })

  describe('posts', () => {
    test('request parameters', async () => {
      mock.onGet('/teams/acme/posts').replyOnce(({ params }) => {
        expect(params).toEqual({
          q: 'text',
          include: 'comments,stargazers',
          sort: 'stars',
          order: 'desc',
          page: 2,
          per_page: 100,
        })
        return [200, {}, {}]
      })
      await api.getPosts({
        q: 'text',
        include: ['comments', 'stargazers'],
        sort: 'stars',
        order: 'desc',
        page: 2,
        per_page: 100,
      })
    })

    test('get 2 posts', async () => {
      mock.onGet('/teams/acme/posts').replyOnce<Partial<PostsPayload>>(
        200,
        {
          posts: [{}, {}] as Post[],
          next_page: 2,
        },
        {},
      )
      const data = await api.getPosts({ per_page: 2 })
      expect(data.posts.length).toBe(2)
    })

    test('pagination', async () => {
      const post1 = { number: 1 } as Post
      mock.onGet('/teams/acme/posts').replyOnce<Partial<PostsPayload>>(
        200,
        {
          posts: [post1] as Post[],
          next_page: 2,
        },
        {},
      )

      const data1 = await api.getPosts({ per_page: 1 })
      expect(data1.posts[0]).toEqual(post1)

      const post2 = { number: 2 } as Post
      mock.onGet('/teams/acme/posts').replyOnce<Partial<PostsPayload>>(
        200,
        {
          posts: [post2] as Post[],
          next_page: 3,
        },
        {},
      )

      const data2 = await api.getPosts({ per_page: 1, page: data1.next_page! })
      expect(data2.posts[0]).toEqual(post2)
    })

    test('embody ratelimit', async () => {
      mock.onGet('/teams/acme/posts').replyOnce<Partial<PostsPayload>>(
        200,
        {
          posts: [],
        },
        {
          'x-ratelimit-limit': '75',
          'x-ratelimit-remaining': '73',
          'x-ratelimit-reset': '1589974200',
        },
      )
      const { ratelimit } = await api.getPosts()
      expect(ratelimit).toEqual({
        limit: 75,
        remaining: 73,
        reset: new Date(1589974200 * 1000),
      })
    })
  })

  test('embody team name', async () => {
    mock.onGet('/teams/acme/posts').replyOnce<Partial<PostsPayload>>(
      200,
      {
        posts: [],
      },
      {},
    )
    const { team } = await api.getPosts()
    expect(team).toBe('acme')
  })
})


gatsby-node.ts
import { SourceNodesArgs } from 'gatsby'
import { Errors } from './lib/errors'
import { EsaApi, PostsPayload } from './lib/esaApi'
import { newEsaNodeCreator } from './lib/esaNodeCreator'
import { isValidPluginOptions, PluginOptions } from './lib/pluginOptions'
import { createSchemaCustomization } from './lib/createSchemaCustomization'

// noinspection JSUnusedGlobalSymbols
export const sourceNodes = async (
  {
    createNodeId,
    createContentDigest,
    actions: { createNode, createParentChildLink },
    getNode,
    reporter,
  }: Pick<
    SourceNodesArgs,
    'createNodeId' | 'createContentDigest' | 'actions' | 'reporter' | 'getNode'
  >,
  options?: PluginOptions,
): Promise<void> => {
  const errors = new Errors()
  if (!isValidPluginOptions(options, errors)) {
    reporter.error(
      `[gatsby-source-esa]\nPlease check your plugin options:\n${errors}`,
    )
    return
  }
  const { team, posts: postsParams = {} } = options
  const esa = new EsaApi({ team, token: options.token })
  const nodeCreator = newEsaNodeCreator({
    team,
    createNode,
    createNodeId,
    createContentDigest,
    getNode,
    createParentChildLink,
  })
  let page: null | number = 1
  while (page) {
    const { posts, next_page }: PostsPayload = await esa.getPosts({
      page,
      ...postsParams,
      per_page: EsaApi.postsPerPageMax,
    })
    await nodeCreator.createPostNodes(posts)
    page = next_page
  }
}

export { createSchemaCustomization }


lib/pluginOptions.ts
import { isObject } from '@suin/is-object'
import { Errors } from './errors'
import { GetPostsParameters } from './esaApi'

export type PluginOptions = {
  /**
   * チーム名
   */
  readonly team: string

  /**
   * esa APIアクセストークン
   */
  readonly token: string

  /**
   * 投稿一覧取得APIのクエリ文字列にセットする情報
   */
  readonly posts?: GetPostsParameters
}

/**
 * @internal
 */
const isPluginOptions = (value: unknown): value is PluginOptions =>
  isObject<PluginOptions>(value) &&
  typeof value.team === 'string' &&
  typeof value.token === 'string'

/**
 * @internal
 */
export const isValidPluginOptions = (
  options: unknown,
  errors: Errors,
): options is PluginOptions => {
  if (!isPluginOptions(options)) {
    errors.add('`team` and `token` property must be specified.')
    return false
  }
  const { team, token } = options
  if (team.length === 0) errors.add('`team` property must be specified.')
  if (token.length === 0) errors.add('`token` property must be specified.')
  return errors.noErrors()
}


lib/esaNodeCreator.ts
import { Actions, NodeInput, NodePluginArgs } from 'gatsby'
import { Post } from './esaApi'

type NewEsaNodeCreator = (params: {
  readonly team: string
  readonly createNode: Actions['createNode']
  readonly createNodeId: NodePluginArgs['createNodeId']
  readonly createContentDigest: NodePluginArgs['createContentDigest']
  readonly getNode: NodePluginArgs['getNode']
  readonly createParentChildLink: Actions['createParentChildLink']
}) => EsaNodeCreator

export interface EsaNodeCreator {
  createPostNodes(posts: ReadonlyArray<Readonly<Post>>): Promise<void>

  createPostNode(post: Readonly<Post>): Promise<string>

  createPostBodyMarkdownNode(
    post: Pick<Readonly<Post>, 'body_md' | 'url'>,
  ): Promise<string>

  createPostBodyHtmlNode(
    post: Pick<Readonly<Post>, 'body_html' | 'url'>,
  ): Promise<string>
}

export interface EsaPost extends Post {
  team: string
}

export const newEsaNodeCreator: NewEsaNodeCreator = ({
  team,
  createNode,
  createNodeId,
  createContentDigest,
  getNode,
  createParentChildLink,
}) => {
  const createPostNode: EsaNodeCreator['createPostNode'] = async (
    post,
    children = [],
  ) => {
    const id = createNodeId(post.url)
    const node: EsaPost & NodeInput = {
      ...post,
      team,
      id,
      internal: {
        type: 'EsaPost',
        contentDigest: createContentDigest(post),
        description: post.url,
      },
      children,
    }
    await createNode(node)
    return id
  }

  const createPostBodyMarkdownNode: EsaNodeCreator['createPostBodyMarkdownNode'] = async ({
    url,
    body_md,
  }) => {
    const id = createNodeId(`${url}.md`)
    await createNode({
      id,
      team: team,
      content: body_md,
      internal: {
        type: 'EsaPostBodyMarkdown',
        contentDigest: createContentDigest(body_md),
        description: `${url}.md`,
        content: body_md,
        mediaType: 'text/markdown',
      },
    })
    return id
  }

  const createPostBodyHtmlNode: EsaNodeCreator['createPostBodyHtmlNode'] = async ({
    body_html,
    url,
  }) => {
    const id = createNodeId(`${url}.html`)
    await createNode({
      id,
      team: team,
      html: body_html, // gatsby-transformer-rehype support
      content: body_html,
      internal: {
        type: 'EsaPostBodyHtml',
        contentDigest: createContentDigest(body_html),
        description: `${url}`,
        content: body_html,
        mediaType: 'text/html',
      },
    })
    return id
  }

  const linkPostChildren = (
    postId: string,
    {
      htmlId,
      markdownId,
    }: {
      readonly htmlId: string
      readonly markdownId: string
    },
  ): void => {
    const markdownNode = getNode(markdownId)
    const htmlNode = getNode(htmlId)
    const postNode = getNode(postId)
    createParentChildLink({ parent: postNode, child: markdownNode })
    createParentChildLink({ parent: postNode, child: htmlNode })
  }

  const createPostNodes: EsaNodeCreator['createPostNodes'] = async posts => {
    for (const post of posts) {
      const markdownId = await createPostBodyMarkdownNode(post)
      const htmlId = await createPostBodyHtmlNode(post)
      const postId = await createPostNode(post)
      linkPostChildren(postId, { htmlId, markdownId })
    }
  }

  return {
    createPostNodes,
    createPostNode,
    createPostBodyHtmlNode,
    createPostBodyMarkdownNode,
  }
}


lib/createSchemaCustomization.ts
import { CreateSchemaCustomizationArgs, GatsbyNode } from 'gatsby'

export const createSchemaCustomization: GatsbyNode['createSchemaCustomization'] = async ({
  actions,
}: CreateSchemaCustomizationArgs): Promise<any> => {
  const { createTypes } = actions
  // language=GraphQL
  const typeDefs = `
    type EsaPost implements Node {
      team: String!
      number: Int!
      name: String!
      full_name: String!
      wip: Boolean!
      body_md: String!
      body_html: String!
      created_at: Date! @dateformat
      message: String!
      kind: String!
      comments_count: Int!
      tasks_count: Int!
      done_tasks_count: Int!
      url: String!
      updated_at: Date! @dateformat
      tags: [String!]!
      category: String!
      revision_number: Int!
      created_by: EsaPostUser!
      updated_by: EsaPostUser!
      stargazers_count: Int!
      watchers_count: Int!
      star: Boolean!
      watch: Boolean!
    }
    
    type EsaPostBodyHtml implements Node {
      team: String!
      html: String!
      content: String!
    }
    
    type EsaPostBodyMarkdown implements Node {
      team: String!
      content: String!
    }
    
    type EsaPostUser {
      name: String!
      screen_name: String!
      icon: String!
    }
  `
  createTypes(typeDefs)
}


lib/esaApi.ts
import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios'

export class EsaApi {
  static readonly postsPerPageMax = 100

  readonly client: AxiosInstance
  private readonly team: string

  constructor({
    team,
    token,
  }: {
    readonly team: string
    readonly token: string
  }) {
    this.team = team
    this.client = axios.create({ baseURL: 'https://api.esa.io/v1' })
    this.client.interceptors.request.use(enheadAuthorization(token))
    this.client.interceptors.response.use(embodyRatelimit)
    this.client.interceptors.response.use(embodyTeamName(team))
  }

  async getPosts({
    q,
    include: includeArray,
    sort,
    order,
    page,
    per_page,
  }: GetPostsParameters & PaginationParameters = {}): Promise<PostsPayload> {
    const include = includeArray ? includeArray.join(',') : undefined
    const { data } = await this.client.request({
      method: 'GET',
      url: `/teams/${this.team}/posts`,
      params: { q, include, sort, order, page, per_page },
    })
    return data
  }
}

/**
 * @internal
 */
const enheadAuthorization = (token: string) => (
  request: AxiosRequestConfig,
): AxiosRequestConfig => {
  request.headers.Authorization = `Bearer ${token}`
  return request
}

/**
 * @internal
 */
const embodyRatelimit = (response: AxiosResponse): AxiosResponse => {
  const limit = parseInt(response.headers['x-ratelimit-limit'])
  const remaining = parseInt(response.headers['x-ratelimit-remaining'])
  const reset = parseInt(response.headers['x-ratelimit-reset'])
  if (
    Number.isInteger(limit) &&
    Number.isInteger(remaining) &&
    Number.isInteger(reset)
  ) {
    response.data.ratelimit = {
      limit,
      remaining,
      reset: new Date(reset * 1000),
    }
  }
  return response
}

/**
 * @internal
 */
const embodyTeamName = (team: string) => (
  response: AxiosResponse,
): AxiosResponse => {
  response.data.team = team
  return response
}

export interface PostsPayload extends PaginatedData {
  posts: Post[]
  ratelimit: Ratelimit
  team: string
}

export interface PaginatedData {
  prev_page: null | number
  next_page: null | number
  total_count: number
  page: number
  per_page: number
  max_per_page: number
}

export interface Post {
  number: number
  name: string
  full_name: string
  wip: boolean
  body_md: string
  body_html: string
  created_at: string
  message: string
  kind: string
  comments_count: number
  tasks_count: number
  done_tasks_count: number
  url: string
  updated_at: string
  tags: string[]
  category: null | string
  revision_number: number
  created_by: User
  updated_by: User
  stargazers_count: number
  watchers_count: number
  star: boolean
  watch: boolean
  sharing_urls: null | {
    html: string
    slides: string
  }
  comments?: Comment[]
  stargazers?: Stargazer[]
}

export interface Comment {
  id: number
  body_md: string
  body_html: string
  created_at: string
  updated_at: string
  post_number: number
  url: string
  created_by: User
  stargazers_count: number
  star: boolean
  stargazers?: Stargazer[]
}

export interface Stargazer {
  created_at: string
  body: null | string
  user: User
}

export interface User {
  name: string
  screen_name: string
  icon: string
}

/**
 * GET /v1/teams/:team_name/postsに指定できるURIクエリ文字列
 *
 * 公式ドキュメント: [dev/esa/api/v1 #noexpand - docs.esa.io](https://docs.esa.io/posts/102#URI%E3%82%AF%E3%82%A8%E3%83%AA%E6%96%87%E5%AD%97%E5%88%97-1)
 */
export interface GetPostsParameters {
  /**
   * 記事を絞り込むための条件を指定します
   *
   * [help/記事の検索方法 - docs.esa.io](https://docs.esa.io/posts/104)を参照
   */
  readonly q?: string

  /**
   * 投稿の関連情報を含めることができます。
   *
   * - `comments` を指定するとコメントの配列を含んだレスポンスを返します。
   * - `stargazers` を指定するとStarの配列を含んだレスポンスを返します。
   * - 複数指定できます。
   */
  readonly include?: ReadonlyArray<
    'comments' | 'comments.stargazers' | 'stargazers'
  >

  /**
   * 投稿の並び順
   *
   * - `updated` (default): 記事の更新日時
   * - `created`: 記事の作成日時
   * - `number`: 記事番号
   * - `stars`: 記事へのStarの数
   * - `watches`: 記事へのWatchの数
   * - `comments`: 記事へのCommentの数
   * - `best_match`: 総合的な記事のスコア
   */
  readonly sort?:
    | 'updated'
    | 'created'
    | 'number'
    | 'stars'
    | 'watches'
    | 'comments'
    | 'best_match'

  /**
   * 並び順の方向
   *
   * - `desc` (default): 降順
   * - `asc`: 昇順
   */
  readonly order?: 'desc' | 'asc'
}

export interface PaginationParameters {
  readonly page?: number
  readonly per_page?: number
}

export interface Ratelimit {
  limit: number
  remaining: number
  reset: Date
}
